daemon/containerd/image_builder.go

```diff
@@ -193,23 +192,55 @@ func newROLayerForImage(ctx context.Context, imgDesc *ocispec.Descriptor, i *Ima
 		return nil, err
 	}
 
+	// TODO(vvoland): Check if image is unpacked, and unpack it if it's not.
 	imageSnapshotID := identity.ChainID(diffIDs).String()
 
+	snapshotter := i.StorageDriver()
+	lease, err := leaseSnapshot(ctx, i.client.LeasesService(), snapshotter, imageSnapshotID)
+	if err != nil {
+		return nil, errdefs.System(fmt.Errorf("failed to lease image snapshot %s: %w", imageSnapshotID, err))
+	}
+
 	return &rolayer{
 		key:                imageSnapshotID,
 		c:                  i.client,
-		snapshotter:        i.snapshotter,
+		snapshotter:        snapshotter,
 		diffID:             "",
 		contentStoreDigest: "",
+		lease:              lease,
 	}, nil
 }
 
+func leaseSnapshot(ctx context.Context, lm leases.Manager, snapshotter string, snapshotID string) (leases.Lease, error) {
+	if snapshotID == "" {
+		return leases.Lease{}, nil
+	}
+
+	id := "legacybuilder_" + stringid.GenerateRandomID()
+	lease, err := lm.Create(ctx, leases.WithID(id), leases.WithExpiration(time.Hour*24))
```

Comment:
I chose name prefix thinking it would be more discoverable when looking at `ctr leases list`, but I see that it also shows labels:

```bash
$ ctr leases ls
ID                                                               CREATED AT           LABELS
bc32e760dbfdd3d6c00f153f362f2695a1514154f5f03ec10a53b8915764171e 2023-09-07T07:33:10Z -
```

So using a label here would also be fine.

Reply:
Changed to label.